tools/cldr-code/src/main/java/org/unicode/cldr/test/RelatedDatePathValues.java

```diff
@@ -0,0 +1,226 @@
+            XPathParts.getFrozenInstance(
+                    "//ldml/dates/calendars/calendar[@type=\"gregorian\"]/dateTimeFormats/intervalFormats/intervalFormatItem[@id=\"Bhm\"]/greatestDifference[@id=\"B\"]");
+
+    // samples
+    // ldml/dates/calendars/calendar[@type="gregorian"]/dateTimeFormats/intervalFormats
+    //  /intervalFormatItem[@id="Bhm"]/greatestDifference[@id="B"]
+    // ldml/dates/calendars/calendar[@type="gregorian"]/dateTimeFormats/availableFormats
+    //  /dateFormatItem[@id="MMMd"]
+    // ldml/dates/calendars/calendar[@type="gregorian"]/dateFormats
+    // /dateFormatLength[@type="full"]/dateFormat[@type="standard"]/datetimeSkeleton[@type="standard"]
+    // ldml/dates/calendars/calendar[@type="gregorian"]/timeFormats
+    // /timeFormatLength[@type="full"]/timeFormat[@type="standard"]/datetimeSkeleton[@type="standard"]
+    public static Set<String> getRelatedPathValues(CLDRFile cldrFile, XPathParts xparts) {
+        if (xparts.size() <= idElement) {
+            return Set.of();
+        }
+        switch (xparts.getElement(dateTypeElement)) {
+            case "availableFormats":
+                return forAvailable(cldrFile, xparts);
+            case "intervalFormats":
+                return forInterval(cldrFile, xparts);
+            default:
+                break;
+        }
+        return Set.of();
+    }
+
+    public enum SkeletonPathType {
+        na,
+        available,
+        interval,
+        datetime;
+
+        public static SkeletonPathType fromParts(XPathParts xparts) {
+            if (xparts.size() >= idElement) {
+                switch (xparts.getElement(dateTypeElement)) {
+                    case "availableFormats":
+                        return xparts.getAttributeValue(idElement, "id") == null ? na : available;
+                    case "intervalFormats":
+                        return xparts.getAttributeValue(idElement, "id") == null ? na : interval;
+                    case "dateFormatLength":
+                    case "timeFormatLength":
+                        return "datetimeSkeleton".equals(xparts.getElement(-1)) ? datetime : na;
+                }
+            }
+            return na;
+        }
+    }
+
+    private static Set<String> forAvailable(CLDRFile cldrFile, XPathParts xparts) {
+        Set<String> skeletons = new LinkedHashSet<>();
+        String skeleton = xparts.getAttributeValue(idElement, "id");
+        addRelated(skeleton, "G", skeletons);
+        addRelated(skeleton, "E", skeletons);
+        addRelated(skeleton, "v", skeletons);
+        if (skeletons.isEmpty()) {
+            return Set.of();
+        }
+        XPathParts newPath = xparts.cloneAsThawed();
+        Set<String> result = new LinkedHashSet<>();
+        for (String item : skeletons) {
+            addAvailable(cldrFile, newPath, item, result);
+        }
+        return result;
+    }
+
+    private static String addRelated(String skeleton, String letter, Set<String> skeletons) {
```

Comment:
addRelated should be void not String; currently it always returns null and the return value is unused